packages/sources/coinmarketcap-test/src/crypto-utils.ts

```diff
@@ -0,0 +1,258 @@
+  const uniqueQuotes = new Set(params.map((p) => p.quote.toUpperCase()))
+  const chunkedMatrix = chunkArray([...uniqueQuotes])
+
+  const groupList = chunkedMatrix
+    .map((chunk) => {
+      const cParams = params.filter((param) => chunk.includes(param.quote.toUpperCase()))
+      return groupByBaseOptions(cParams)
+    })
+    .flat()
+
+  return groupList.map((group) => {
+    const queryName = group[0]
+    const groupedParams = group[1]
+
+    return {
+      params: groupedParams.map((inputParams) => inputParams.payload),
+      request: {
+        baseURL: config.API_ENDPOINT,
+        url: '/cryptocurrency/quotes/latest',
+        headers: {
+          'X-CMC_PRO_API_KEY': config.API_KEY,
+        },
+        params: {
+          [queryName]: [
+            ...new Set(
+              groupedParams.map((p) => p.id || p.payload.cid || p.payload.slug || p.payload.base),
+            ),
+          ].join(','),
+          convert: [...new Set(groupedParams.map((p) => p.payload.quote))].join(','),
+        },
+      },
+    }
+  })
+}
+
+// CMC API currently uses ID as key in response, when querying with "slug" param
+const _keyForSlug = (data: ProviderResponseBody, slug: string) => {
+  if (!data || !data.data) return
+  // First try to find slug key in response (in case the API starts returning keys correctly)
+  if (Object.keys(data.data).includes(slug)) return slug
+  // Fallback to ID
+  const _iEqual = (s1: string, s2: string) => s1.toUpperCase() === s2.toUpperCase()
+  const o = Object.values(data.data).find((o) => _iEqual(o.slug, slug))
+
+  return o && o.id
+}
+
+// If there is a base-to-id override in presetId file we need to get the id of coin because that will be in the response
+const _keyForSymbol = (data: ProviderResponseBody, symbol: string) => {
+  if (data.data[symbol.toUpperCase()]) {
+    return symbol.toUpperCase()
+  }
+  if ((presetIds as Record<string, number>)[symbol.toUpperCase()]) {
+    return (presetIds as Record<string, number>)[symbol.toUpperCase()].toString()
+  }
+
+  return symbol.toUpperCase()
+}
+
+export const constructEntry = (
+  params: CryptoRequestParams[],
+  response: ProviderResponseBody,
+  resultPath: keyof PriceInfo,
+) => {
+  return params.map((param) => {
+    const quote = param.quote.toUpperCase()
+
+    let key: string | number | undefined = param.cid
+    if (!key && param.slug) {
+      key = _keyForSlug(response, param.slug)
+    }
+    if (!key) {
+      key = _keyForSymbol(response, param.base)
+    }
+
+    if (!response.data[key]) {
+      return errorResponse(param, `Data for "${param.cid || param.slug || param.base}" not found`)
+    }
+
+    const dataForQuote = response.data[key].quote[quote]
+    if (!dataForQuote) {
+      return errorResponse(
+        param,
+        `"${dataForQuote}" quote for "${param.cid || param.slug || param.base}" not found`,
```

Comment:
Wouldn't `dataForQuote` be undefined here? Should this be just `quote`?

Reply:
Yep, correct